### PR TITLE
ci: Add public_interface command

### PR DESCRIPTION
**Pull request by aahung**
### Description of changes

This command allows extraction of all public interfaces into a file. This allows us to compare two files to determine if there is a compatibility-breaking changes.

### Checklist

- [ ] Adheres to the [development guidelines](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines)
- [ ] Add/update [transform tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions)
    - [ ] Using correct values
    - [ ] Using wrong values
- [ ] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)

### Examples?

```
bin/public_interface.py extract out.json
```

Snippet of generated JSON file:
```
{
  "constants": [
    "samtranslator.feature_toggle.feature_toggle.BOTO3_CONNECT_TIMEOUT",
    "samtranslator.intrinsics.resolver.DEFAULT_SUPPORTED_INTRINSICS",
    "samtranslator.model.api.api_generator.GatewayResponseProperties",
    "samtranslator.model.api.http_api_generator.DefaultStageName",
    "samtranslator.model.api.http_api_generator.HttpApiTagName",
    "samtranslator.model.apigatewayv2.APIGATEWAY_AUTHORIZER_KEY",
    "samtranslator.model.architecture.ARM64",
    "samtranslator.model.architecture.X86_64",
    "samtranslator.model.connector_profiles.profile.PROFILE",
    "samtranslator.model.connector_profiles.profile.f",
    "samtranslator.model.eventsources.FUNCTION_EVETSOURCE_METRIC_PREFIX",
    "samtranslator.model.eventsources.cloudwatchlogs.FUNCTION_EVETSOURCE_METRIC_PREFIX",
    "samtranslator.model.eventsources.pull.FUNCTION_EVETSOURCE_METRIC_PREFIX",
    "samtranslator.model.eventsources.push.CONDITION",
    "samtranslator.model.eventsources.push.FUNCTION_EVETSOURCE_METRIC_PREFIX",
    "samtranslator.model.eventsources.push.REQUEST_PARAMETER_PROPERTIES",
    "samtranslator.model.eventsources.scheduler.FUNCTION_EVETSOURCE_METRIC_PREFIX",
    "samtranslator.model.intrinsics.MIN_NUM_CONDITIONS_TO_COMBINE",
    "samtranslator.model.packagetype.IMAGE",
    "samtranslator.model.packagetype.ZIP",
    "samtranslator.model.preferences.deployment_preference_collection.CODEDEPLOY_APPLICATION_LOGICAL_ID",
    "samtranslator.model.preferences.deployment_preference_collection.CODEDEPLOY_PREDEFINED_CONFIGURATIONS_LIST",
    "samtranslator.model.preferences.deployment_preference_collection.CODE_DEPLOY_CONDITION_NAME",
    "samtranslator.model.preferences.deployment_preference_collection.CODE_DEPLOY_SERVICE_ROLE_LOGICAL_ID",
    "samtranslator.model.sam_resources.ARM64",
    "samtranslator.model.sam_resources.IMAGE",
    "samtranslator.model.sam_resources.X86_64",
    "samtranslator.model.sam_resources.ZIP",
    "samtranslator.model.stepfunctions.events.CONDITION",
    "samtranslator.model.stepfunctions.events.SFN_EVETSOURCE_METRIC_PREFIX",
    "samtranslator.plugins.api.implicit_api_plugin.MIN_NUM_CONDITIONS_TO_COMBINE",
    "samtranslator.plugins.application.serverless_app_plugin.BOTO3_CONNECT_TIMEOUT",
    "samtranslator.plugins.application.serverless_app_plugin.PLUGIN_METRICS_PREFIX",
    "samtranslator.policy_template_processor.template.POLICY_PARAMETER_DISAMBIGUATE_PREFIX",
    "samtranslator.policy_templates_data.POLICY_TEMPLATES_FILE",
    "samtranslator.policy_templates_data.SCHEMA_FILE",
    "samtranslator.translator.translator.TYPE_CHECKING",
    "samtranslator.translator.verify_logical_id.do_not_verify",
    "samtranslator.utils.constants.BOTO3_CONNECT_TIMEOUT",
    "samtranslator.utils.py27hash_fix.MINSIZE",
    "samtranslator.utils.py27hash_fix.PERTURB_SHIFT",
    "samtranslator.validator.sam_schema.SCHEMA_DIR",
    "samtranslator.validator.sam_schema.SCHEMA_FILE",
    "samtranslator.validator.sam_schema.SCHEMA_NEW_FILE",
    "samtranslator.validator.validator.INTRINSIC_ATTR"
  ],
  "routines": {
    "samtranslator.feature_toggle.dialup.BaseDialup.is_enabled": [
      {
        "default": null,
        "kind": "POSITIONAL_OR_KEYWORD",
        "name": "self"
      }
    ],
    "samtranslator.feature_toggle.dialup.DisabledDialup.is_enabled": [
      {
        "default": null,
        "kind": "POSITIONAL_OR_KEYWORD",
        "name": "self"
      }
    ],
    "samtranslator.feature_toggle.dialup.SimpleAccountPercentileDialup.is_enabled": [
      {
        "default": null,
        "kind": "POSITIONAL_OR_KEYWORD",
        "name": "self"
      }
    ],
    "samtranslator.feature_toggle.dialup.ToggleDialup.is_enabled": [
      {
        "default": null,
        "kind": "POSITIONAL_OR_KEYWORD",
        "name": "self"
      }
    ],
    "samtranslator.feature_toggle.feature_toggle.FeatureToggle.is_enabled": [
      {
        "default": null,
        "kind": "POSITIONAL_OR_KEYWORD",
        "name": "self"
      },
      {
        "default": null,
        "kind": "POSITIONAL_OR_KEYWORD",
        "name": "feature_name"
      }
    ],
    "samtranslator.intrinsics.actions.Action.can_handle": [
      {
        "default": null,
        "kind": "POSITIONAL_OR_KEYWORD",
        "name": "self"
      },
      {
        "default": null,
        "kind": "POSITIONAL_OR_KEYWORD",
        "name": "input_dict"
      }
    ],
    "samtranslator.intrinsics.actions.Action.resolve_parameter_refs": [
      {
        "default": null,
        "kind": "POSITIONAL_OR_KEYWORD",
        "name": "self"
      },
      {
        "default": null,
        "kind": "POSITIONAL_OR_KEYWORD",
        "name": "input_dict"
      },
      {
        "default": null,
        "kind": "POSITIONAL_OR_KEYWORD",
        "name": "parameters"
      }
    ],
    "samtranslator.intrinsics.actions.Action.resolve_resource_id_refs": [
      {
        "default": null,
        "kind": "POSITIONAL_OR_KEYWORD",
        "name": "self"
      },
      {
        "default": null,
        "kind": "POSITIONAL_OR_KEYWORD",
        "name": "input_dict"
      },
      {
```

Please reach out in the comments if you want to add an example. Examples will be 
added to `sam init` through [aws/aws-sam-cli-app-templates](https://github.com/aws/aws-sam-cli-app-templates).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by acristin — APPROVED** [review]
lgtm; is the plan to diff with previous release?

**Comment by aahung** [comment]
> lgtm; is the plan to diff with previous release?

Diff with the base branch. And it is not a simple diff, the new one should be able to compatible with the old one:
- can add constants or routines
- can add new optional arguments.

**Review by ssenchenko — APPROVED** [review]

File: bin/public_interface.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    extract.add_argument(
+        "output_file",
+        help="The path to output JSON file",
+        type=Path,
+    )
+    args = parser.parse_args()
+
+    if args.command == "extract":
+        scanner = InterfaceScanner()
+        scanner.scan_interfaces_recursively("samtranslator")
```
Comment (hoffa) [inline bin/public_interface.py:100]: Can you make this an optional argument? So it's generic and usable wherever.

Comment (aahung) [reply bin/public_interface.py:100]: Good idea

File: bin/public_interface.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        for function_name, function in inspect.getmembers(
+            importlib.import_module(module_name), lambda obj: inspect.ismethod(obj) or inspect.isfunction(obj)
+        ):
+            # Skip imported functions and ones starting with "_"
+            if function.__module__ != module_name or function_name.startswith("_"):
+                continue
+            full_path = f"{module_name}.{function_name}"
+            self.signatures[full_path] = inspect.signature(function)
+
+    def _scan_constants_in_module(self, module_name: str) -> None:
```
Comment (hoffa) [inline bin/public_interface.py:39]: It's not necessarily constants though. Nor do I think it matters. These are module-scoped variables, we don't know if they mutate. Maybe just `variables`?

Comment (aahung) [reply bin/public_interface.py]: updated

File: bin/public_interface.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
+#!/usr/bin/env python
+"""Extract/compare public interfaces."""
```
Comment (hoffa) [inline bin/public_interface.py:2]: Should have a blurb about why this exists in the first place. Maybe here.

Something like "transform library hasn't had a publicly defined interface, so we assume anything not considered private by convention (see https://peps.python.org/pep-0008/#descriptive-naming-styles) is public and needs to maintain backward compatibility; this helps automate such checks"

Comment (aahung) [reply bin/public_interface.py]: added something

File: bin/public_interface.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        json.dump(result, f, indent=2, sort_keys=True)
+
+
+def main() -> None:
+    parser = argparse.ArgumentParser(description=__doc__)
+
+    subparsers = parser.add_subparsers(dest="command")
+    extract = subparsers.add_parser("extract", help="Extract public interfaces into a JSON file")
+    extract.add_argument("--module", help="The module to extract public interfaces", type=str, default="samtranslator")
+    extract.add_argument(
```
Comment (hoffa) [inline bin/public_interface.py:102]: For consideration, printing to `stdout` might be easier. It's more versatile as it's trivial to pipe to another file, and also allows inline diffs such as `diff <(bin/public_interface.py samtranslator) old-interface.json`.

**Review by hoffa — APPROVED** [review]

File: bin/public_interface.py (hunk 4 above)
Comment (aahung) [reply bin/public_interface.py:102]: good idea but normal people comment in batch

**Pull request merged**